# BioData Catalyst Tutorial

This Dockerfile provides a toy example to demonstrate how to [Bring Your Own Tool](https://bdcatalyst.gitbook.io/biodata-catalyst-documentation/written-documentation/community-tools-and-integration-1/bring-your-own-tool-s-to-biodata-catalyst). The script is simple and takes an RNA sequence and converts it to an amino acid sequence. 

## Overview

**What is BioData Catalyst**

The BDC ecosystem offers secure workspaces to support your data analysis in addition to a number of bioinformatics tools for analysis. 

**What is this tutorial for**

The NHLBI BioData Catalyst Data Management Center helps investigators bring their data to BioData Catalyst and use the platform. The [Navigating BioData Catalyst YouTube series](https://www.youtube.com/playlist?list=PLSIGopyrrU80U9oytmP6vbkOdtimi5FAb) offers short form tutorials to provide step-by-step instruction for using the platform. 

**What is BYOT**
BDC supports lots of different data types but not all. In some cases, datatypes may not have available tooling to help analyze the data. In this case, users can bring dockerized tools and workflows to the platform to aid in data analysis. 

<br>

## Usage
This is a simple script that translates a default or user supplied RNA sequence into an amino-acid sequence. 

The following command can be used to run the docker: 
```
docker pull rtibiocloud/bdc_tutorial:<tagname>
docker run -it rtibiocloud/bdc_tutorial:<tagname> -c "Rscript /opt/bdc_tutorial/aa_translator.R --help"
```

It is recommended that the latest version is used for most applications:
```
rtibiocloud/bdc_tutorial:v1.0_9b4d5cd
```

<br>


## Build
To build this Docker image, you can use the following command:
```
docker build --rm -t rtibiocloud/bdc_tutorial:<tagname> -f Dockerfile .`
```
Here's what each part of the command does:

`docker build`: This command tells Docker to build an image.
`--rm`: This flag removes any intermediate containers that are created during the build process, helping to keep your system clean.
`-t bdc_tutorial:v1.0.0`: The -t flag specifies the name and tag for the image. In this case, it's named bdc_tutorial with version v1.0.0.
`-f Dockerfile`: This flag specifies the Dockerfile to use for building the image. You can replace Dockerfile with the actual name of your Dockerfile if it's different.
`.`: The dot at the end of the command indicates that the build context is the current directory, where the Dockerfile is located.
Running this command will build a Docker image with the name `bdc_tutorial:v1.0.0`. Make sure you are in the directory containing the Dockerfile you want to use for building the image.

## Perform a testrun
`docker run -it 69d989ee9140 bdc_tutorial testrun --id=check_install`

## Rscript Inputs
| Short Flag | Long Flag | Type | Description | Default Value |
|:-----:|:--------:|:-------------:|--------------------------------|---------------|
|   -i  |  --input     | String | RNA sequence to translate to amino acids             | 'CAUAUCUAA' |
|   -o  |  --outfile   | String | File name for the output file                        | 'tutorial_output.txt' |
|   -3  |  --three     | Flag   | Display amino acids using three letter abbreviations | FALSE | 
|   -w  |  --write     | Flag   | Write the translated sequence to file                | FALSE |
|   -v  |  --verbose   | Flag   | Display verbose logging                              | FALSE | 
|   -h  |  --help      | Flag   | Display the function usage statement                 | FALSE |

## Rscript Output
### Translated Amino Acid Sequence
If the --write flag was supplied the script will also output a .txt file with the translated amino acid sequence. By default, this output will use the single letter amino acid symbols and will be named tutorial_output.tsv, however, users can specify a file name using the --outfile flag

An example output input RNA sequence AGUUACAUAUGCAAA: 
```
STICK
```

<br>

If the --three flag was used, the output will be a hyphen delimiated sequence using the three letter amino acid abbreviations:
```
Ser-Thr-Ile-Cys-Lys
```
### Log File
The Rscript will produce a log file describing the environment and the commands execution. The log file will have the same basename as outputs, but with a .log extension. 
A tab-delimited log file with 4 columns and a variable number of rows, depending on if the --verbose flag is provided.
|     Column     | Description                                            |
|:--------------:|--------------------------------------------------------|
|   Timestamp    | Time stamp when the message was logged                 |
|     Level      | Level of the message: info, debug, warning, error      |
|   Function     | The function that created the logging statment         |
|    Message     | The message generated by the script                    |

An example log file would look like this
```
timestamp	                level	function    message
[2025-06-13 18:47:59.185195]	info	main	    User: root
[2025-06-13 18:47:59.207768]	info	main	    Running from: c7494424390b
[2025-06-13 18:47:59.208641]	info	main	    Platform: x86_64-pc-linux-gnu
[2025-06-13 18:47:59.212934]	info	main	    R version: R version 4.4.2 (2024-10-31)
[2025-06-13 18:47:59.213711]	info	main	    R packages loaded: getopt
[2025-06-13 18:47:59.216216]	info	main	    Rscript: /opt/bdc_tutorial/aa_translator.R
[2025-06-13 18:47:59.21702]	info	getopt	    CommandLine: -i AUGAUGAUGGUAGUAGUA
[2025-06-13 18:47:59.217784]	info	getopt	    Arguments: ARGS = character(0)
[2025-06-13 18:47:59.217784]	info	getopt	    Arguments: input = AUGAUGAUGGUAGUAGUA
[2025-06-13 18:47:59.217784]	info	getopt	    Arguments: three = FALSE
[2025-06-13 18:47:59.217784]	info	getopt	    Arguments: verbose = FALSE
[2025-06-13 18:47:59.217784]	info	getopt	    Arguments: write = FALSE
[2025-06-13 18:47:59.217784]	info	getopt	    Arguments: outfile = tutorial_output.txt
[2025-06-13 18:47:59.21864]	info	main	    Current Working Directory: /sbgenomics/workspaces/fc3bdcb7-47fa-4d16-ad3a-f7f4f3038d0e/tasks/a282fc30-60fb-4e0b-bce5-9d45ce50da8a/rmip-analysis-tool-example
[2025-06-13 18:47:59.237575]	info	main	    The RNA sequence AUGAUGAUGGUAGUAGUA translates to the protein sequence: MMMGGG
[2025-06-13 18:47:59.238592]	info	main	    Process began at 2025-06-13 18:47:59.173216 and finished at 2025-06-13 18:47:59.23871
```

## Perform a testrun
`docker run -v ${PWD}/example_files/:/data -t rtibiocloud/bdc_tutorial:v1.0.0 /bin/bash  -c "Rscript /opt/bdc_tutorial/aa_translator.R -i CAUAUCUAA -3 -v -w"`

<details>

```
Loading required package: getopt
Warning message:
package 'getopt' was built under R version 4.4.1
[2025-06-11 16:11:03.385816] - main - INFO - User: jstratford
[2025-06-11 16:11:03.401193] - main - INFO - Running from: RTI-110812
[2025-06-11 16:11:03.403926] - main - INFO - Platform: x86_64-w64-mingw32/x64
[2025-06-11 16:11:03.409529] - main - INFO - R version: R version 4.4.0 (2024-04-24 ucrt)
[2025-06-11 16:11:03.411584] - main - INFO - R packages loaded: getopt
[2025-06-11 16:11:03.415139] - main - INFO - Rscript: bdc_tutorial.R
[2025-06-11 16:11:03.417573] - getopt - INFO - CommandLine: -w -v -i CAUAUCUAA -3
[2025-06-11 16:11:03.420022] - getopt - INFO - Arguments: ARGS = character(0)
[2025-06-11 16:11:03.420022] - getopt - INFO - Arguments: write = TRUE
[2025-06-11 16:11:03.420022] - getopt - INFO - Arguments: verbose = TRUE
[2025-06-11 16:11:03.420022] - getopt - INFO - Arguments: input = CAUAUCUAA
[2025-06-11 16:11:03.420022] - getopt - INFO - Arguments: three = TRUE
[2025-06-11 16:11:03.420022] - getopt - INFO - Arguments: outfile = tutorial_output.txt
[2025-06-11 16:11:03.428271] - main - INFO - Current Working Directory: C:/Users/jstratford/Documents/github/biocloud_docker_tools/bdc_tutorials/v1.0
[2025-06-11 16:11:03.439941] - main - INFO - The RNA sequence CAUAUCUAA translates to the protein sequence: His-Ile-STOP
[2025-06-11 16:11:03.445898] - main - INFO - Process began at 2025-06-11 16:11:03.366598 and finished at 2025-06-11 16:11:03.446056
[2025-06-11 16:11:03.451276] - export - INFO - Translation exported to tutorial_output.txt
[2025-06-11 16:11:03.461224] - main - INFO - Finished
```

<br>

```
Outputs:
- Tab separated table:                      /data/tutorial_output.txt
- Tab separated log file:                   /data/tutorial_output.log
```
</details>

<br>

## Contact
For additional information or assistance, please contact Jeran Stratford (jstratford@rti.org).
